### PR TITLE
File Parsing Updates

### PR DESCRIPTION
**Pull request by gabergg**

### CODE REVIEW

File: desktop/src/utils/asyncFs.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import fs from 'fs'
```
Comment (gdowens) [inline desktop/src/utils/asyncFs.js]: Nearly certain this file isn't used, if so we should just delete it.

Comment (gabergg) [inline desktop/src/utils/asyncFs.js]: yep, i thought the same. wasnt sure if there was a pending purpose

File: desktop/libs/Project/decoentry.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// modify this variable to move around components and screens
+let Entry = require('./index')
+
+// point to index if this is a release build
+if (!__DEV__) {
+  Entry = require('./index')
+}
+
+// Check if we're dealing with ES6-style imports
+if (typeof Entry === 'object' && Entry !== null && Entry.default) {
```
Comment (dabbott) [inline desktop/libs/Project/decoentry.js]: I guess in practice it shouldn't matter... but technically shouldn't we check for Entry.__esModule here?

File: desktop/libs/Project/index.android.js [hunk 3]
```diff
@@ -49,4 +48,4 @@ const styles = StyleSheet.create({
   },
 });
 
-AppRegistry.registerComponent('Project', () => Project);
+export default Project
```
Comment (dabbott) [inline desktop/libs/Project/index.android.js]: We're gonna duck punch `AppRegistry.registerComponent` instead at some point?

File: desktop/libs/Project/ios/Project/AppDelegate.m [hunk 4]
```diff
@@ -31,7 +31,7 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
    * on the same Wi-Fi network.
    */
 
-  jsCodeLocation = [NSURL URLWithString:@"http://localhost:8081/index.ios.bundle?platform=ios&dev=true"];
+  jsCodeLocation = [NSURL URLWithString:@"http://localhost:8081/decoentry.bundle?platform=ios&dev=true"];
```
Comment (dabbott) [inline desktop/libs/Project/ios/Project/AppDelegate.m]: Wonder if there's something less intrusive we can do here

File: desktop/src/utils/codemod/expressionMethods.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
+import j from 'jscodeshift'
```
Comment (dabbott) [inline desktop/src/utils/codemod/expressionMethods.js]: Would be nice to have these codemod-related files on web instead of desktop.

I don't know all the pros & cons, but big pro: don't have to restart anything... can probably hot reload, too.

File: web/src/scripts/actions/storyboardActions.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  TOGGLE_VIEW: 'TOGGLE_VIEW',
+}
+
+export const toggleStoryboardView = () => async (dispatch) => {
+  dispatch({
+    type: at.TOGGLE_VIEW,
+  })
+}
+
+export const openStoryboard = (filepath) => async (dispatch, getState) => {
```
Comment (dabbott) [inline web/src/scripts/actions/storyboardActions.js]: Wonder if we can break this up into smaller functions

File: web/src/scripts/actions/storyboardActions.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    payload: {
+      connections: sceneConnections,
+      scenes: scenes,
+      entry: sceneInfo.entry,
+    },
+  })
+}
+
+export const addScene = () => async (dispatch, getState) => {
+  const scaffoldId = 0 // id corresponds to Component
```
Comment (dabbott) [inline web/src/scripts/actions/storyboardActions.js]: Probably shouldn't check for id 0. We can make an API to get by name or something. Or change ids to `com.decosoftware.storyboard` or something, idk

File: web/src/scripts/containers/Storyboard.jsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+import { createSelector } from 'reselect'
+import { StylesEnhancer } from 'react-styles-provider'
+import YOPS from 'yops'
+
+import * as ContentLoader from '../api/ContentLoader'
+import * as URIUtils from '../utils/URIUtils'
+import { storyboardActions } from '../actions'
+import NewSceneButton from '../components/storyboard/NewSceneButton'
+
+const stylesCreator = (theme) => {
```
Comment (dabbott) [inline web/src/scripts/containers/Storyboard.jsx]: Not important, but normally I write these in the more compact form

```
const stylesCreator = ({colors}) => ({
  container: {
    backgroundColor: 'white',
    flex: '1 1 auto',
  }
})
```

File: web/src/scripts/containers/TabbedEditor.jsx [hunk 9]
```diff
@@ -252,6 +256,37 @@ class TabbedEditor extends Component {
     this.setState({showMenu: false})
   }
 
+  renderEditor = () => {
```
Comment (dabbott) [inline web/src/scripts/containers/TabbedEditor.jsx]: Should delete this. It isn't called. It was moved into the Editor container

File: web/src/scripts/reducers/storyboardReducer.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  shouldShow: false,
+}
+
+export default (state = initialState, action) => {
+  const {type, payload} = action
+
+  switch(type) {
+    case at.OPEN_STORYBOARD: {
+      return update(state, {
+        connections: {$set: payload.connections},
```
Comment (dabbott) [inline web/src/scripts/reducers/storyboardReducer.js]: Not important, but I guess use `$merge` here, since 3 properties

File: web/src/scripts/utils/storyboard/codeUtils.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
+import _ from 'lodash'
+import ElementTreeBuilder from '../ElementTreeBuilder'
+const CodeMod = Electron.remote.require('./utils/codemod/index.js')
+const path = Electron.remote.require('path')
+
+const formatFilePaths = (filepath, options = {}) => {
```
Comment (dabbott) [inline web/src/scripts/utils/storyboard/codeUtils.js]: Needs comment

File: web/src/scripts/utils/storyboard/codeUtils.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        sceneName: _.get(sceneImport, 'identifiers[0].value'),
+        source: formattedPath,
+      }
+    })
+  return scenes
+}
+
+/**
+ * Takes thes scene code and looks for the push and the pops
+ * @param  {[type]} code [description]
```
Comment (dabbott) [inline web/src/scripts/utils/storyboard/codeUtils.js]: Delete or fill out boilerplate types

File: web/src/scripts/utils/storyboard/codeUtils.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  }
+}
+
+export const buildElementTree = (code) => {
+  const elementTree = ElementTreeBuilder.elementTreeFromAST(CodeMod(code).nodes())
+}
+
+/**
+ * Returns an object of scenes, sceneNames, entryScene, whatever
+ * @param  {[type]} code [description]
```
Comment (dabbott) [inline web/src/scripts/utils/storyboard/codeUtils.js]: Delete or update types

File: web/src/styles/main.scss [hunk 14]
```diff
@@ -455,6 +455,19 @@ $icons-list:insert save logout settings to-code to-smart-object record-stop reco
...
 
+.icon-play {
+	&:after {
+		content: "►";
+		width: 14px;
+		height: 22px;
+		color: rgba(0,0,0,0.50);
+		font-size: 14px;
+		position: relative;
+		// top: -2px;
```
Comment (dabbott) [inline web/src/styles/main.scss]: Delete comment

File: desktop/src/utils/codemod/expressionMethods.js (hunk 5 above)
Comment (gabergg) [inline desktop/src/utils/codemod/expressionMethods.js]: remote require jscodeshift from web?

File: desktop/src/utils/codemod/expressionMethods.js [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        j.identifier(arg.property)
+      )
+    }
+    default: {
+      return j.literal(arg.value || "")
+    }
+  }
+})
+
+const inversecreateArgumentNodes = (args = []) => args.map((arg) => {
```
Comment (gabergg) [inline desktop/src/utils/codemod/expressionMethods.js]: camel

File: desktop/src/utils/codemod/expressionMethods.js [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+/**
+ * Removes all functions that exactly match the description of 'object.property(...args)'
+ * If args is undefined, it matches only on 'object.property'
+ *
+ * @param Same as addFunctionCall
+ * @return Same as addFunctionCall
+ */
+export const removeFunctionCall = function(object, property, args) {
+  const program = _.get(this.nodes(), '[0].program', {})
```
Comment (gabergg) [inline desktop/src/utils/codemod/expressionMethods.js]: fix for export case?

Comment (gabergg) [inline desktop/src/utils/codemod/expressionMethods.js]: TEST

File: web/src/scripts/actions/storyboardActions.js [hunk 17]
```diff
@@ -0,0 +1,182 @@
+/**
+ *    Copyright (C) 2015 Deco Software Inat.
```
Comment (gabergg) [inline web/src/scripts/actions/storyboardActions.js]: critical bug

File: web/src/scripts/components/filetree/Node.js [hunk 18]
```diff
@@ -87,17 +87,44 @@ export default class Node extends Component {
...
+    if (!(menuVisible || hover || pinned)) {
       return null
     }
 
+    if (!isDirectory(type)) {
+      return (
+        <div style={styles.plusContainer} onClick={this.stopPropagation}>
+          <PlayButton
+            node={node}
+            active={pinned}
```
Comment (gabergg) [inline web/src/scripts/components/filetree/Node.js]: fix pin state

File: web/src/scripts/components/filetree/PlayButton.jsx [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import TwoStateButton from '../buttons/TwoStateButton'
+
+export default ({
+  node, onChange, active
+}) => {
+  return (
+    <TwoStateButton
+      enabled={active}
+      onChange={onChange}
+      hoverStyle={{ opacity: 0.75 }}
```
Comment (gabergg) [inline web/src/scripts/components/filetree/PlayButton.jsx]: pull out?

File: web/src/scripts/factories/scaffold/ComponentScaffolds/Scene.js [hunk 20]
```diff
@@ -0,0 +1,51 @@
+/**
```
Comment (gabergg) [inline web/src/scripts/factories/scaffold/ComponentScaffolds/Scene.js]: maybe use this over component for scaffold?

File: web/src/scripts/factories/storyboard/EntryFileChangeFactory.js [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ */
+
+import _ from 'lodash'
+
+import DecoChangeFactory from '../editor/DecoChangeFactory'
+const CodeMod = Electron.remote.require('./utils/codemod/index.js')
+
+class EntryFileChangeFactory {
+  static createChangeToUpdateEntryRequire(doc, relativePath) {
```
Comment (gabergg) [inline web/src/scripts/factories/storyboard/EntryFileChangeFactory.js]: generally, descriptive doc naming

File: web/src/scripts/factories/storyboard/index.js [hunk 22]
```diff
@@ -0,0 +1,2 @@
+export { default as EntryFileChangeFactory } from './EntryFileChangeFactory'
```
Comment (gabergg) [inline web/src/scripts/factories/storyboard/index.js]: succinct

File: web/src/scripts/actions/storyboardActions.js [hunk 23]
```diff
@@ -42,66 +42,86 @@ export const toggleStoryboardView = () => async (dispatch) => {
...
       sceneName,
     }
   })
+}
 
-  // match source info to component layout info
-  // return storyboard object
+export const openStoryboard = (filepath) => async (dispatch, getState) => {
+  const rootPath = getState().directory.rootPath
+  console.log(getState().directory)
```
Comment (dabbott) [inline web/src/scripts/actions/storyboardActions.js]: console.log

File: web/src/scripts/containers/ProjectNavigator.jsx [hunk 24]
```diff
@@ -104,17 +108,20 @@ class ProjectNavigator extends Component {
...
   handlePreviewClick = async (node) => {
     const {rootPath, textEditorCompositeActions} = this.props
+    this.setState({currentPreview: node.path})
     const relativePath = node.path.replace(rootPath, '.')
     const requireText = relativePath.replace(path.extname(relativePath), '')
     textEditorCompositeActions.updateDecoEntryRequire(requireText)
   }
 
-  renderNode = (props) => {
+  renderNode = (props, state) => {
```
Comment (dabbott) [inline web/src/scripts/containers/ProjectNavigator.jsx]: probably you didn't mean to pass in this state, cause state isn't passed to this function, and you're using this.state below

Comment (gabergg) [inline web/src/scripts/containers/ProjectNavigator.jsx]: yuuup true

File: web/src/scripts/containers/ProjectNavigator.jsx [hunk 25]
```diff
@@ -104,17 +108,20 @@ class ProjectNavigator extends Component {
 
   handlePreviewClick = async (node) => {
     const {rootPath, textEditorCompositeActions} = this.props
+    this.setState({currentPreview: node.path})
```
Comment (dabbott) [inline web/src/scripts/containers/ProjectNavigator.jsx]: I think this could go in another PR, but really the play button is determined by the entry point, which we pull out of a file. So wherever we have that in the store, we should pass that through to here and there should be a redux action to set it

Comment (gdowens) [inline web/src/scripts/containers/ProjectNavigator.jsx]: +1

**Comment by dabbott** [comment]
👍

File: web/src/scripts/containers/ProjectNavigator.jsx (hunk 25 above)
Comment (gabergg) [inline web/src/scripts/containers/ProjectNavigator.jsx]: Yep, planning to do that in the next PR. Wasn't sure if this was the final interaction or anything, and the redux store for this stuff is still in motion. Will improve it.

**Comment by gdowens** [comment]
I think I'm using approve wrong lol
